### ISSUE TITLE
Possibly quadratic behaviour

### ISSUE THREAD

**Issue by nad** [agda/agda#5983]
I grepped for `\+\+ \[` and found several pieces of code that can perhaps lead to quadratic behaviour (I ignored code under `Agda.Auto`):

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Compiler/MAlonzo/Compiler.hs#L150

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Interaction/Options/Base.hs#L782

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Abstract/Views.hs#L49

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Concrete.hs#L621

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Concrete/Operators.hs#L864

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Parser/Parser.y#L579

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Parser/Parser.y#L1348

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Translation/AbstractToConcrete.hs#L402

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Translation/AbstractToConcrete.hs#L855

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/TypeChecking/Monad/Caching.hs#L78

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/TypeChecking/Records.hs#L883

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/TypeChecking/Records.hs#L892

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/TypeChecking/Rules/LHS.hs#L981

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/TypeChecking/Rules/LHS.hs#L1144

https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/TypeChecking/Substitute.hs#L121-L122

**Comment by andreasabel** [comment]
Thanks for investigating.  The individual items would be more valuable with reproducers (although I think you couldn't pin performance problems to just one of these items, esp. when the problem is triggered by large terms).

I guess the most relevant of these would be concerning _lists of eliminations in internal syntax_.  We typically read these first-to-last but often need to append a new elimination at the end.  The quadratic factor would matter if you have long argument lists or deep nesting of projections.
A relevant experiment would be to replace the _list_ by `Data.Sequence` or something similar and see if overall performance increases or decreases.

**Comment by nad** [comment]
> A relevant experiment would be to replace the _list_ by `Data.Sequence` or something similar and see if overall performance increases or decreases.

Given the high "constant factors" that I think are associated with `Data.Sequence` you might see both better and worse performance, depending on the test case.

**Comment by jespercockx** [comment]
A more lightweight alternative might be to use difference lists? (e.g. https://hackage.haskell.org/package/dlist-1.0/docs/Data-DList.html)

**Comment by andreasabel** [comment]
Not sure about `dlist`.  They are good for construction, but as soon as you start consuming, they need to be collapsed into ordinary lists.  If you store a `dlist`, then every inspection needs to do the collapsing again.

**Comment by nad** [comment]
> We typically read these first-to-last but often need to append a new elimination at the end.

How are these lists used? Would a FIFO queue data structure suffice, or some kind of deque, or are elements also accessed based on their position?

**Comment by UlfNorell** [comment]
When projecting from a constructor application you need the *i*th argument in the list of eliminators.

**Comment by andreasabel** [comment]
(Removing label "bug": There is no evidence for such claim here.)

**Comment by nad** [comment]
> (Removing label "bug": There is no evidence for such claim here.)

Here is one example. Compare Agda's performance for the following families of files:
```agda
{-# FOREIGN GHC
    {-# #-} {-# #-} {-# #-} {-# #-} {-# #-} {-# #-} {-# #-} {-# #-}
⋮
    {-# #-} {-# #-} {-# #-} {-# #-} {-# #-} {-# #-} {-# #-} {-# #-}
  #-}
```
```agda
{-# FOREIGN GHC
    {-# {-# {-# {-# {-# {-# {-# {-# {-# {-# {-# {-# {-# {-# {-# {-#
⋮
    #-} #-} #-} #-} #-} #-} #-} #-} #-} #-} #-} #-} #-} #-} #-} #-}
  #-}
```

**Comment by andreasabel** [comment]
I didn't do the "families", but both these instances parse instantly.

**Comment by nad** [comment]
Try with hundreds or thousands of lines instead of `⋮`.

**Comment by UlfNorell** [comment]
Can you first explain why we care about efficient parsing of pragmas nested thousands of levels deep?

**Comment by nad** [comment]
I don't think this is very important, but Andreas asked for evidence of quadratic behaviour.

Here is a possibly more relevant example of superlinear performance:
```agda
postulate
  F : Set₁ → … → Set₁

_ = F Set … Set
```
When I doubled the number of arguments (from 130 to 260, and then to 520) the type-checking time increased by first a factor of ∼5, and then a factor of ∼6, to about six seconds (on my machine). I don't know why Agda is so slow, but I can imagine that there are several reasons.

I also tried similar Haskell code:
```haskell
f :: Int -> … -> Int -> Int
f _ … _ = 0

n = f 0 … 0
```
GHCi quickly loaded this code (with 520 arguments).
